### PR TITLE
NATS binding API and refactor transport

### PR DESCRIPTION
**Pull request by dan-j**
As per #317 and some discussion in the slack channel

I've also gone and refactored `nats.Transport`.

Everything should be implemented but I've not added tests yet... just wanted some feedback that I'm on the right track before spending time on tests

### CODE REVIEW

**Comment by n3wscott** [comment]
Ah sorry, stuff is moving around, can you rebase?

**Comment by dan-j** [comment]
Sure thing.. still need to write tests though 😁

File: pkg/bindings/nats/encoder.go [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+	if err != nil {
+		return err
+	}
+
+	e.Data = bytes
+
+	return nil
+}
+
+func (e *natsMsgEncoder) Start(ctx context.Context) error {
```
Comment (slinkydeveloper) [inline pkg/bindings/nats/encoder.go]: You don't need to implement the `BinaryEncoder` methods at all

File: pkg/bindings/nats/errors.go [hunk 2]
```diff
@@ -0,0 +1,5 @@
+package nats
+
+import "fmt"
+
+var BinaryEncodingNotSupported = fmt.Errorf("binary encoding for NATS protocol is not supported")
```
Comment (slinkydeveloper) [inline pkg/bindings/nats/errors.go]: You don't need it, if you need to signal that you don't support the binary encoding, you must use `binding.ErrNotBinary`

File: pkg/bindings/nats/message.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+}
+
+var _ binding.Message = (*Message)(nil)
+
+// By the specification, all NATS message payloads must be in the JSON event format.
+// If this isn't the case then this function will return an error
+func NewMessage(msg *nats.Msg) (*Message, error) {
+	raw := make(map[string]json.RawMessage)
+
+	if err := json.Unmarshal(msg.Data, &raw); err != nil {
```
Comment (slinkydeveloper) [inline pkg/bindings/nats/message.go]: Why do you perform this unmarshalling?

**Comment by dan-j** [comment]
Sorry I've been quiet a couple of weeks, but I'm back on this (and STAN) now...

Is this still required? It looks like the latest v2 updates addressed this?

**Comment by n3wscott** [comment]
Yes, sorry, we worked through nats because we needed to drop the codec and the transport interfaces.

I think this is now implemented.

**Comment by slinkydeveloper** [comment]
@n3wscott is this one still relevant? can we close it?

**Pull request closed without merging**